diff --git a/arch/x86/platform/geode/alix.c b/arch/x86/platform/geode/alix.c
index 90e23e7..8197530 100644
--- a/arch/x86/platform/geode/alix.c
+++ b/arch/x86/platform/geode/alix.c
@@ -43,7 +43,7 @@ MODULE_PARM_DESC(force, "Force detection as ALIX.2/ALIX.3 platform");
 static struct gpio_keys_button alix_gpio_buttons[] = {
 	{
 		.code			= KEY_RESTART,
-		.gpio			= 24,
+		.gpio			= 1,
 		.active_low		= 1,
 		.desc			= "Reset button",
 		.type			= EV_KEY,
