--- gst-plugins-base-0.10.35/gst/typefind/gsttypefindfunctions.c.orig	2011-06-14 14:52:17.000000000 +0200
+++ gst-plugins-base-0.10.35/gst/typefind/gsttypefindfunctions.c	2011-09-26 22:44:08.000000000 +0200
@@ -4327,8 +4327,10 @@
       mpeg4_video_type_find, m4v_exts, MPEG_VIDEO_CAPS, NULL, NULL);
   TYPE_FIND_REGISTER (plugin, "video/x-h263", GST_RANK_SECONDARY,
       h263_video_type_find, h263_exts, H263_VIDEO_CAPS, NULL, NULL);
+#if 0
   TYPE_FIND_REGISTER (plugin, "video/x-h264", GST_RANK_PRIMARY,
       h264_video_type_find, h264_exts, H264_VIDEO_CAPS, NULL, NULL);
+#endif
   TYPE_FIND_REGISTER (plugin, "video/x-nuv", GST_RANK_SECONDARY, nuv_type_find,
       nuv_exts, NUV_CAPS, NULL, NULL);
 
