diff -Naur gst-plugins-base-0 orig.10.25/gst/playback/gstdecodebin.c gst-plugins-base-0.10.25/gst/playback/gstdecodebin.c
--- gst-plugins-base-0 orig.10.25/gst/playback/gstdecodebin.c	2009-12-06 14:45:04.000000000 +0100
+++ gst-plugins-base-0.10.25/gst/playback/gstdecodebin.c	2009-12-06 14:33:21.000000000 +0100
@@ -1012,6 +1012,13 @@
 static gboolean
 is_demuxer_element (GstElement * srcelement)
 {
+  const char *srcName = gst_element_get_name(srcelement);
+
+  if(strstr(srcName, "oggdemux") == srcName)
+    {
+      return FALSE;
+    }
+
   GstElementFactory *srcfactory;
   GstElementClass *elemclass;
   GList *walk;
