--- uClibc-0.9.32/libc/inet/getaddrinfo.c.orig	2011-07-11 14:04:51.000000000 +0200
+++ uClibc-0.9.32/libc/inet/getaddrinfo.c	2011-07-11 14:05:52.000000000 +0200
@@ -177,8 +177,16 @@
 			seen |= SEEN_IPV4;
 #endif
 #if defined __UCLIBC_HAS_IPV6__
-		if (runp->ifa_addr->sa_family == PF_INET6)
+		if (runp->ifa_addr->sa_family == PF_INET6) {
+			const struct in6_addr *addr = &((struct sockaddr_in6 *) runp->ifa_addr)->sin6_addr;
+			if (IN6_IS_ADDR_LOOPBACK(addr))
+				continue;
+			if (IN6_IS_ADDR_LINKLOCAL(addr))
+				continue;
+			if (IN6_IS_ADDR_MC_LINKLOCAL(addr))
+				continue;
 			seen |= SEEN_IPV6;
+		}
 #endif
 	}
 	freeifaddrs(ifa);
