=== modified file 'Makefile'
--- Makefile	2010-11-18 20:28:16 +0000
+++ Makefile	2011-06-27 15:47:12 +0000
@@ -41,18 +41,17 @@
 
 ifdef ARCH
 
-CC =  $(CROSS_STAGING_DIR)/usr/bin/$(ARCH)-linux-gcc
-CXX = $(CROSS_STAGING_DIR)/usr/bin/$(ARCH)-linux-g++
-LD =  $(CROSS_STAGING_DIR)/usr/bin/$(ARCH)-linux-gcc
-
 CROSS_PREFIX      = $(HOME)/buildroot
-CROSS_STAGING_DIR = $(CROSS_PREFIX)/build_$(ARCH)/staging_dir
-CROSS_HOST_DIR    = $(CROSS_PREFIX)/build_$(ARCH)/host_dir
+CROSS_STAGING_DIR = $(CROSS_PREFIX)/output/staging
+CROSS_HOST_DIR    = $(CROSS_PREFIX)/output/host
 CROSS_INSTALL_DIR = $(HOME)/root.$(ARCH)/raumfeld/p0-renderer
 
-DBUS_BINDING_TOOL = $(CROSS_PREFIX)/build_arm/host_dir/usr/bin/dbus-binding-tool
+CC =  $(CROSS_HOST_DIR)/usr/bin/$(ARCH)-linux-gcc
+CXX = $(CROSS_HOST_DIR)/usr/bin/$(ARCH)-linux-g++
+LD =  $(CROSS_HOST_DIR)/usr/bin/$(ARCH)-linux-gcc
 
-GLIB_GENMARSHAL   = $(CROSS_PREFIX)/build_$(ARCH)/host_dir/usr/bin/glib-genmarshal
+DBUS_BINDING_TOOL = $(CROSS_HOST_DIR)/usr/bin/dbus-binding-tool
+GLIB_GENMARSHAL   = $(CROSS_HOST_DIR)/usr/bin/glib-genmarshal
 
 PKG_CONFIG             = $(CROSS_HOST_DIR)/usr/bin/pkg-config
 PKG_CONFIG_SYSROOT_DIR = $(CROSS_STAGING_DIR)

=== modified file 'test/Makefile'
--- test/Makefile	2010-09-08 22:40:16 +0000
+++ test/Makefile	2011-06-27 15:47:35 +0000
@@ -16,18 +16,18 @@
 		$(OBJDIR)/p0-eq.o
 
 ifndef CROSS_PREFIX
-CROSS_PREFIX      = $(HOME)/buildroot
+CROSS_PREFIX = $(HOME)/buildroot
 endif
 
-PKG_CONFIG      = pkg-config
+PKG_CONFIG   = pkg-config
 
 ifeq ($(CROSS),ARM)
-ARCH=arm
 
-CROSS_STAGING_DIR = $(CROSS_PREFIX)/build_arm/staging_dir
+CROSS_STAGING_DIR = $(CROSS_PREFIX)/output/staging
+CROSS_HOST_DIR    = $(CROSS_PREFIX)/output/host
 CROSS_INSTALL_DIR = $(HOME)/root.arm/raumfeld/p0-audiotest
 
-PKG_CONFIG             = $(CROSS_PREFIX)/build_arm/host_dir/usr/bin/pkg-config
+PKG_CONFIG             = $(CROSS_HOST_DIR)/usr/bin/pkg-config
 PKG_CONFIG_SYSROOT_DIR = $(CROSS_STAGING_DIR)
 PKG_CONFIG_PATH        = $(CROSS_STAGING_DIR)/usr/lib/pkgconfig
 
@@ -40,8 +40,8 @@
 	--sysroot=$(CROSS_STAGING_DIR) \
 	-isysroot $(CROSS_STAGING_DIR)
 
-CC = $(CROSS_STAGING_DIR)/usr/bin/arm-linux-gcc
-LD = $(CROSS_STAGING_DIR)/usr/bin/arm-linux-gcc
+CC = $(CROSS_HOST_DIR)/usr/bin/arm-linux-gcc
+LD = $(CROSS_HOST_DIR)/usr/bin/arm-linux-gcc
 
 INSTALLDIR = $(CROSS_INSTALL_DIR)
 INSTALLFILE = $(APP)

