--- gst-plugins-ugly-0.10.10/ext/mad/gstmad.c.orig	2008-11-08 15:48:46.000000000 +0100
+++ gst-plugins-ugly-0.10.10/ext/mad/gstmad.c	2009-03-24 12:45:19.000000000 +0100
@@ -1621,7 +1621,14 @@ gst_mad_chain (GstPad * pad, GstBuffer *
           goto skip_frame;
         }
 
-        mad_synth_frame (&mad->synth, &mad->frame);
+        if (mad->stream.error == MAD_ERROR_BADDATAPTR) {
+          mad->stream.error = MAD_ERROR_NONE;
+          mad_frame_mute (&mad->frame);
+          mad_synth_mute (&mad->synth);
+        } else {
+          mad_synth_frame (&mad->synth, &mad->frame);
+        }
+
         left_ch = mad->synth.pcm.samples[0];
         right_ch = mad->synth.pcm.samples[1];
 
