--- faad2-2.6.1/libfaad/syntax.c	2007-11-01 13:33:40.000000000 +0100
+++ faad2-2.6.1-fixed/libfaad/syntax.c	2009-11-05 10:51:32.000000000 +0100
@@ -433,6 +433,12 @@
         while ((id_syn_ele = (uint8_t)faad_getbits(ld, LEN_SE_ID
             DEBUGVAR(1,4,"NeAACDecDecode(): id_syn_ele"))) != ID_END)
         {
+          if(id_syn_ele == 0 && ld->bytes_left == 0)
+          {
+            printf("skipping SCE frame\n");
+            break;
+          }
+
             switch (id_syn_ele) {
             case ID_SCE:
                 ele_this_frame++;
