From 69cc4871c2492430eb17af3fda73c3e400c3e7a0 Mon Sep 17 00:00:00 2001
From: Sven Neumann <neumann@teufel.de>
Date: Mon, 25 Nov 2013 16:05:23 +0100
Subject: [PATCH] Skip overly strict verification of DIDL-Lite XML

---
 libgupnp-av/gupnp-didl-lite-parser.c | 9 ---------
 1 file changed, 9 deletions(-)

diff --git a/libgupnp-av/gupnp-didl-lite-parser.c b/libgupnp-av/gupnp-didl-lite-parser.c
index 086c2cc..73fdf60 100644
--- a/libgupnp-av/gupnp-didl-lite-parser.c
+++ b/libgupnp-av/gupnp-didl-lite-parser.c
@@ -374,15 +374,6 @@ parse_elements (GUPnPDIDLLiteParser *parser,
                         }
                 } else if (GUPNP_IS_DIDL_LITE_ITEM (object)) {
                         node = gupnp_didl_lite_object_get_xml_node (object);
-                        if (!verify_didl_attributes (node)) {
-                                g_object_unref (object);
-                                g_set_error (error,
-                                             GUPNP_XML_ERROR,
-                                             GUPNP_XML_ERROR_INVALID_ATTRIBUTE,
-                                             "Could not parse DIDL-Lite XML");
-
-                                return FALSE;
-                        }
 
                         g_signal_emit (parser,
                                        signals[ITEM_AVAILABLE],
-- 
1.8.3.2

