=== modified file 'Makefile'
--- Makefile	2010-10-11 14:15:51 +0000
+++ Makefile	2011-06-28 13:19:06 +0000
@@ -16,14 +16,14 @@
 
 ifdef ARCH
 
-CC =  $(CROSS_STAGING_DIR)/usr/bin/$(ARCH)-linux-gcc
-LD =  $(CROSS_STAGING_DIR)/usr/bin/$(ARCH)-linux-gcc
-
 CROSS_PREFIX      = $(HOME)/buildroot
-CROSS_STAGING_DIR = $(CROSS_PREFIX)/build_$(ARCH)/staging_dir
-CROSS_HOST_DIR    = $(CROSS_PREFIX)/build_$(ARCH)/host_dir
+CROSS_STAGING_DIR = $(CROSS_PREFIX)/staging
+CROSS_HOST_DIR    = $(CROSS_PREFIX)/host
 CROSS_INSTALL_DIR = $(HOME)/root.$(ARCH)/raumfeld/report-daemon
 
+CC =  $(CROSS_HOST_DIR)/usr/bin/$(ARCH)-linux-gcc
+LD =  $(CROSS_HOST_DIR)/usr/bin/$(ARCH)-linux-gcc
+
 PKG_CONFIG             = $(CROSS_HOST_DIR)/usr/bin/pkg-config
 PKG_CONFIG_SYSROOT_DIR = $(CROSS_STAGING_DIR)
 PKG_CONFIG_PATH        = $(CROSS_STAGING_DIR)/usr/lib/pkgconfig

