From ee92d0f8143c7838d1b8d5f6974a0317ad8ec2e0 Mon Sep 17 00:00:00 2001
From: Max Blagay <max.blagaj@gmail.com>
Date: Wed, 19 Nov 2014 16:44:01 +0100
Subject: [PATCH] Added Makefile for libasf

---
 .../Makefile                                       | 64 ++++++++++++++++++++++
 1 file changed, 64 insertions(+)
 create mode 100644 Makefile

diff --git a/Makefile b/Makefile
new file mode 100644
index 0000000..ef9619f
--- /dev/null
+++ b/Makefile
@@ -0,0 +1,69 @@
+
+GCC ?= gcc
+DESTDIR ?= /usr/local
+
+OBJDIR = obj
+
+LIBNAME = libasf
+
+HEADERS=\
+	src/asf.h \
+	src/asfint.h \
+	src/byteio.h \
+	src/compat.h \
+	src/data.h \
+	src/debug.h \
+	src/guid.h \
+	src/header.h \
+	src/parse.h \
+	src/utf.h
+
+SRC =\
+	src/asf.c \
+	src/byteio.c \
+	src/data.c \
+	src/guid.c \
+	src/header.c \
+	src/parse.c
+
+OBJS= $(patsubst %.c, $(OBJDIR)/%.o, $(SRC))
+
+COMPILE_C = @echo "compiling $@";\
+	$(GCC) -Isrc -c -MD -o $@ $< $(CFLAGS);
+
+
+all: MKDIR $(LIBNAME)
+
+
+$(LIBNAME): $(OBJS)
+	@echo "link static library $@"
+	@$(AR) crs $(OBJDIR)/$(LIBNAME).a $(OBJS)
+
+
+$(OBJDIR)/%.o: %.c
+	$(COMPILE_C)
+
+
+MKDIR:
+	@echo "Making directories..."
+	@mkdir -p $(OBJDIR)/src
+
+
+clean:
+	rm -f $(OBJS)
+	rm -f $(OBJDIR)/$(LIBNAME).a
+	rm -rf $(OBJDIR)
+
+
+install: all
+	mkdir -p $(DESTDIR)/lib/
+	mkdir -p $(DESTDIR)/include/asf/
+	cp $(OBJDIR)/$(LIBNAME).a $(DESTDIR)/lib/
+	cp -r $(HEADERS) $(DESTDIR)/include/asf/
+
+
+install-exec: all
+	mkdir -p $(DESTDIR)/lib/
+	cp $(OBJDIR)/$(LIBNAME).a $(DESTDIR)/lib/
+
+
-- 
1.9.1

