--- libglib2-2.31.18-orig/gio/Makefile.am	2012-02-20 21:23:17.000000000 +0100
+++ libglib2-2.31.18/gio/Makefile.am	2012-02-29 12:10:35.680474728 +0100
@@ -647,6 +647,7 @@
 glib_compile_resources_LDADD = \
 	$(top_builddir)/glib/libglib-2.0.la \
 	$(top_builddir)/gobject/libgobject-2.0.la	\
+	$(top_builddir)/gmodule/libgmodule-2.0.la	\
 	libgio-2.0.la
 
 glib_compile_resources_SOURCES = \
@@ -677,6 +678,7 @@
 gsettings_LDADD = \
 	$(top_builddir)/glib/libglib-2.0.la		\
 	$(top_builddir)/gobject/libgobject-2.0.la	\
+	$(top_builddir)/gmodule/libgmodule-2.0.la	\
 	libgio-2.0.la
 gsettings_SOURCES = gsettings-tool.c
 
@@ -690,7 +692,8 @@
 gdbus_SOURCES = gdbus-tool.c
 gdbus_LDADD = libgio-2.0.la \
 	$(top_builddir)/glib/libglib-2.0.la 		\
-	$(top_builddir)/gobject/libgobject-2.0.la
+	$(top_builddir)/gobject/libgobject-2.0.la	\
+	$(top_builddir)/gmodule/libgmodule-2.0.la
 
 completiondir = $(sysconfdir)/bash_completion.d
 completion_DATA = \
@@ -707,6 +710,7 @@
 gresource_LDADD = libgio-2.0.la \
 	$(top_builddir)/glib/libglib-2.0.la 		\
 	$(top_builddir)/gobject/libgobject-2.0.la \
+	$(top_builddir)/gmodule/libgmodule-2.0.la	\
 	$(ELF_LIBS)
 
 # ------------------------------------------------------------------------
