From ea2c89c8b7e69fe81b8c9ca8f0e3ab33a54e88e1 Mon Sep 17 00:00:00 2001
From: Slobodan Vukanovic <slobodan.vukanovic@teufel.de>
Date: Thu, 7 Jan 2016 14:59:55 +0100
Subject: [PATCH] C++ strings stay in scope for g_assert_cmpstr

--- a/glib/gtestutils.h
+++ b/glib/gtestutils.h
@@ -39,10 +39,12 @@
 
 /* assertion API */
 #define g_assert_cmpstr(s1, cmp, s2)    G_STMT_START { \
-                                             const char *__s1 = (s1), *__s2 = (s2); \
+                                             char *__s1 = g_strdup(s1), *__s2 = g_strdup(s2); \
                                              if (g_strcmp0 (__s1, __s2) cmp 0) ; else \
                                                g_assertion_message_cmpstr (G_LOG_DOMAIN, __FILE__, __LINE__, G_STRFUNC, \
                                                  #s1 " " #cmp " " #s2, __s1, #cmp, __s2); \
+                                             g_free(__s2); \
+                                             g_free(__s1); \
                                         } G_STMT_END
 #define g_assert_cmpint(n1, cmp, n2)    G_STMT_START { \
                                              gint64 __n1 = (n1), __n2 = (n2); \
