--- libmad-0.15.1b.orig/frame.c	2004-02-04 23:59:19.000000000 +0100
+++ libmad-0.15.1b/frame.c	2009-03-09 18:04:08.000000000 +0100
@@ -450,11 +450,15 @@
 
   frame->header.flags &= ~MAD_FLAG_INCOMPLETE;
 
-  if (decoder_table[frame->header.layer - 1](stream, frame) == -1) {
-    if (!MAD_RECOVERABLE(stream->error))
-      stream->next_frame = stream->this_frame;
+  if (decoder_table[frame->header.layer - 1](stream, frame) == -1)
+  {
+    if(stream->error != MAD_ERROR_BADDATAPTR)
+    {
+      if (!MAD_RECOVERABLE(stream->error))
+        stream->next_frame = stream->this_frame;
 
-    goto fail;
+      goto fail;
+    }
   }
 
   /* ancillary_data() */
