--- gst-ffmpeg-0.10.10/gst-libs/ext/ffmpeg/libavcodec/alac.c	2010-03-06 14:11:26.000000000 +0100
+++ gst-ffmpeg-0.10.10/gst-libs/ext/ffmpeg/libavcodec/alac.c	2010-08-30 16:09:30.000000000 +0200
@@ -132,8 +132,7 @@
     bytestream_get_be16(&ptr);      /* ??? */
     bytestream_get_be32(&ptr);      /* max coded frame size */
     bytestream_get_be32(&ptr);      /* bitrate ? */
-    bytestream_get_be32(&ptr);      /* samplerate */
-
+    alac->avctx->sample_rate = bytestream_get_be32(&ptr);      /* samplerate */
     allocate_buffers(alac);
 
     return 0;
