From 7ce792d56eba6c4d0ef0f175f482da256568a952 Mon Sep 17 00:00:00 2001
From: Oliver Tappe <zooey@hirschkaefer.de>
Date: Tue, 10 Mar 2015 14:17:56 +0100
Subject: [PATCH 2/3] Accept empty passphrases.

---
 src/technology.c | 3 ---
 1 file changed, 3 deletions(-)

diff --git a/src/technology.c b/src/technology.c
index 175c778..8c8b349 100644
--- a/src/technology.c
+++ b/src/technology.c
@@ -985,9 +985,6 @@ static DBusMessage *set_property(DBusConnection *conn,
 		if (technology->type != CONNMAN_SERVICE_TYPE_WIFI)
 			return __connman_error_not_supported(msg);
 
-		if (strlen(str) < 8 || strlen(str) > 63)
-			return __connman_error_passphrase_required(msg);
-
 		if (g_strcmp0(technology->tethering_passphrase, str) != 0) {
 			g_free(technology->tethering_passphrase);
 			technology->tethering_passphrase = g_strdup(str);
-- 
1.8.4.5

