diff --git a/directfb.pc.in b/directfb.pc.in
index 7e53453..af6f907 100644
--- a/directfb.pc.in
+++ b/directfb.pc.in
@@ -7,6 +7,6 @@ Name: DirectFB
 Description: Graphics and windowing library for the Linux frame buffer device
 Version: @VERSION@
 Requires: @DEP_VOODOO@ fusion direct
-Libs: -L${libdir} -ldirectfb @THREADLIB@ @OSX_LIBS@
+Libs: -L${libdir} -ldirectfb @THREADLIB@ @OSX_LIBS@ -lm
 Libs.private: -L${libdir} @DYNLIB@ @ZLIB_LIBS@
 Cflags: @THREADFLAGS@ -I@INCLUDEDIR@
