diff --git a/systems/fbdev/fbdev_surface_pool.c b/systems/fbdev/fbdev_surface_pool.c
index a129771..2f597a4 100644
--- a/systems/fbdev/fbdev_surface_pool.c
+++ b/systems/fbdev/fbdev_surface_pool.c
@@ -322,6 +322,29 @@ fbdevDeallocateBuffer( CoreSurfacePool       *pool,
 }
 
 static DFBResult
+fbdevMuckOut( CoreSurfacePool   *pool,
+              void              *pool_data,
+              void              *pool_local,
+              CoreSurfaceBuffer *buffer )
+{
+     CoreSurface        *surface;
+     FBDevPoolData      *data  = pool_data;
+     FBDevPoolLocalData *local = pool_local;
+
+     D_DEBUG_AT( FBDev_Surfaces, "%s( %p )\n", __FUNCTION__, buffer );
+
+     D_MAGIC_ASSERT( pool, CoreSurfacePool );
+     D_MAGIC_ASSERT( data, FBDevPoolData );
+     D_MAGIC_ASSERT( local, FBDevPoolLocalData );
+     D_MAGIC_ASSERT( buffer, CoreSurfaceBuffer );
+
+     surface = buffer->surface;
+     D_MAGIC_ASSERT( surface, CoreSurface );
+
+     return dfb_surfacemanager_displace( local->core, data->manager, buffer );
+}
+
+static DFBResult
 fbdevLock( CoreSurfacePool       *pool,
            void                  *pool_data,
            void                  *pool_local,
@@ -406,6 +429,8 @@ const SurfacePoolFuncs fbdevSurfacePoolFuncs = {
      .AllocateBuffer     = fbdevAllocateBuffer,
      .DeallocateBuffer   = fbdevDeallocateBuffer,
 
+     .MuckOut            = fbdevMuckOut,
+
      .Lock               = fbdevLock,
      .Unlock             = fbdevUnlock,
 };
