diff --git a/directfb.pc.in b/directfb.pc.in
index 7e53453..aea8021 100644
--- a/directfb.pc.in
+++ b/directfb.pc.in
@@ -8,5 +8,5 @@ Description: Graphics and windowing library for the Linux frame buffer device
 Version: @VERSION@
 Requires: @DEP_VOODOO@ fusion direct
 Libs: -L${libdir} -ldirectfb @THREADLIB@ @OSX_LIBS@
-Libs.private: -L${libdir} @DYNLIB@ @ZLIB_LIBS@
+Libs.private: -L${libdir} @MEDIALIB@ @DYNLIB@ @ZLIB_LIBS@
 Cflags: @THREADFLAGS@ -I@INCLUDEDIR@
