--- uClibc-0.9.30.1-orig/libc/inet/getaddrinfo.c	2009-02-26 13:49:14.000000000 +0100
+++ uClibc-0.9.30.1/libc/inet/getaddrinfo.c	2010-05-20 20:57:02.000000000 +0200
@@ -193,7 +193,8 @@
 #endif /* __UCLIBC_HAS_IPV4__ */
 #if defined __UCLIBC_HAS_IPV6__
 			if (runp->ifa_addr->sa_family == PF_INET6)
-				seen |= SEEN_IPV6;
+				if (! IN6_IS_ADDR_LINKLOCAL (runp->ifa_addr))
+					seen |= SEEN_IPV6;
 #endif /* __UCLIBC_HAS_IPV6__ */
 		}
 		freeifaddrs(ifa);
