diff -Naur uCLibc-orig/libpthread/linuxthreads/specific.c uClibc-0.9.30.1/libpthread/linuxthreads/specific.c
--- uCLibc-orig/libpthread/linuxthreads/specific.c	2009-09-25 16:42:32.000000000 +0200
+++ uClibc-0.9.30.1/libpthread/linuxthreads/specific.c	2009-09-25 16:43:35.000000000 +0200
@@ -110,7 +110,7 @@
   struct pthread_key_delete_helper_args args;
   args.idx1st = key / PTHREAD_KEY_2NDLEVEL_SIZE;
   args.idx2nd = key % PTHREAD_KEY_2NDLEVEL_SIZE;
-  if (__pthread_manager_request != -1)
+  if (__pthread_manager_request != -1 &&  __pthread_exit_requested == 0)
     {
       struct pthread_request request;
 
