--- uClibc-0.9.30.1-orig/libc/inet/getaddrinfo.c	2010-05-26 13:25:31.000000000 +0200
+++ uClibc-0.9.30.1/libc/inet/getaddrinfo.c	2010-05-26 13:22:34.000000000 +0200
@@ -194,8 +194,16 @@
 				seen |= SEEN_IPV4;
 #endif /* __UCLIBC_HAS_IPV4__ */
 #if defined __UCLIBC_HAS_IPV6__
-			if (runp->ifa_addr->sa_family == PF_INET6)
-				seen |= SEEN_IPV6;
+			if (runp->ifa_addr->sa_family == PF_INET6) {
+                                const struct in6_addr *addr = &((struct sockaddr_in6 *) runp->ifa_addr)->sin6_addr;
+                                if (IN6_IS_ADDR_LOOPBACK(addr))
+                                        continue;
+                                if (IN6_IS_ADDR_LINKLOCAL(addr))
+                                        continue;
+                                if (IN6_IS_ADDR_MC_LINKLOCAL(addr))
+                                        continue;
+                                seen |= SEEN_IPV6;
+                        }
 #endif /* __UCLIBC_HAS_IPV6__ */
 		}
 		freeifaddrs(ifa);
