diff --git a/configure.in b/configure.in
index c17b9da..0971bcb 100644
--- a/configure.in
+++ b/configure.in
@@ -276,6 +276,7 @@ PKG_MAINTAINER=pkg_maintainer
 old_LIBS="$LIBS"
 LIBS=""
 AC_SEARCH_LIBS(backtrace, execinfo)
+AM_CONDITIONAL(HAVE_BACKTRACE, test "x$ac_cv_search_backtrace" != xno)
 BACKTRACELIB="$LIBS"
 LIBS="$old_LIBS"
 
diff --git a/tests/Makefile.am b/tests/Makefile.am
index 59ad3b3..8a298ca 100644
--- a/tests/Makefile.am
+++ b/tests/Makefile.am
@@ -46,9 +46,12 @@ endif #ARCH_IA64
 			test-async-sig test-flush-cache test-init-remote \
 			test-mem test-setjmp test-ptrace		 \
 			Ltest-nomalloc Ltest-nocalloc rs-race
- noinst_PROGRAMS_cdep = forker mapper test-ptrace-misc test-varargs	\
-			Gperf-simple Lperf-simple			\
-			Gperf-trace Lperf-trace
+ noinst_PROGRAMS_cdep = forker mapper test-ptrace-misc			 \
+			Gperf-simple Lperf-simple
+
+if HAVE_BACKTRACE
+ noinst_PROGRAMS_cdep += Gperf-trace Lperf-trace test-varargs
+endif
 
 if SUPPORT_CXX_EXCEPTIONS
  check_PROGRAMS_cdep += Ltest-cxx-exceptions
