--- taglib-1.7.2/taglib/toolkit/tbytevector.cpp.orig	2012-09-28 15:17:31.852984027 +0200
+++ taglib-1.7.2/taglib/toolkit/tbytevector.cpp	2012-09-28 15:17:40.820984031 +0200
@@ -468,7 +468,7 @@
   {
     // we already know the resulting buffers size, so allocate it
     size_t newBuffersSize = size() + (withSize - patternSize) * numPatternPositions;
-    ByteVector result(newBuffersSize);
+    ByteVector result((uint) newBuffersSize);
 
     size_t readPos = 0;
     size_t writePos = 0;
