--- /home/hhoegelo/buildroot/build_i586/taglib-1.5/taglib/ogg/oggpage.cpp	2008-02-04 16:14:46.000000000 +0100
+++ ./taglib/ogg/oggpage.cpp	2010-04-14 10:47:33.000000000 +0200
@@ -107,23 +107,28 @@
   if(index == lastPacketIndex)
     flags = ContainsPacketFlags(flags | EndsWithPacket);
 
-  // If there's only one page and it's complete:
-
-  if(packetCount() == 1 &&
-     !d->header.firstPacketContinued() &&
-     d->header.lastPacketCompleted())
+  // peek into the next page to see if the current
+  // packet is continued there
+  Page nextPage(d->file, fileOffset() + size());
+  if(!nextPage.d->header.isValid() || !nextPage.d->header.firstPacketContinued())
   {
-    flags = ContainsPacketFlags(flags | CompletePacket);
-  }
+    // If there's only one page and it's complete:
+
+    if (packetCount() == 1 && !d->header.firstPacketContinued()
+        && d->header.lastPacketCompleted())
+    {
+      flags = ContainsPacketFlags(flags | CompletePacket);
+    }
 
-  // Or if the page is (a) the first page and it's complete or (b) the last page
-  // and it's complete or (c) a page in the middle.
+    // Or if the page is (a) the first page and it's complete or (b) the last page
+    // and it's complete or (c) a page in the middle.
 
-  else if((flags & BeginsWithPacket && !d->header.firstPacketContinued()) ||
-          (flags & EndsWithPacket && d->header.lastPacketCompleted()) ||
-          (!(flags & BeginsWithPacket) && !(flags & EndsWithPacket)))
-  {
-    flags = ContainsPacketFlags(flags | CompletePacket);
+    else if ((flags & BeginsWithPacket && !d->header.firstPacketContinued())
+        || (flags & EndsWithPacket && d->header.lastPacketCompleted())
+        || (!(flags & BeginsWithPacket) && !(flags & EndsWithPacket)))
+    {
+      flags = ContainsPacketFlags(flags | CompletePacket);
+    }
   }
 
   return flags;
