--- dhcp-4.1-ESV-R8/omapip/dispatch.c.orig	2013-01-03 00:25:28.000000000 +0100
+++ dhcp-4.1-ESV-R8/omapip/dispatch.c	2013-08-23 16:23:50.603315377 +0200
@@ -47,6 +47,16 @@
 OMAPI_OBJECT_ALLOC (omapi_waiter,
 		    omapi_waiter_object_t, omapi_type_waiter)
 
+static void get_monotonic_time (struct timeval *tv)
+{
+  struct timespec tp;
+
+  clock_gettime (CLOCK_MONOTONIC, &tp);
+
+  tv->tv_sec  = tp.tv_sec;
+  tv->tv_usec = tp.tv_nsec / 1000;
+}
+
 void
 register_eventhandler(struct eventqueue **queue, void (*handler)(void *))
 {
@@ -332,7 +342,7 @@
 
 	/* First, see if the timeout has expired, and if so return. */
 	if (t) {
-		gettimeofday (&now, (struct timezone *)0);
+                get_monotonic_time (&now);
 		cur_tv.tv_sec = now.tv_sec;
 		cur_tv.tv_usec = now.tv_usec;
 		if (now.tv_sec > t -> tv_sec ||
@@ -410,7 +420,7 @@
 	}
 
 	/* Get the current time... */
-	gettimeofday (&cur_tv, (struct timezone *)0);
+        get_monotonic_time (&cur_tv);
 
 	/* We probably have a bad file descriptor.   Figure out which one.
 	   When we find it, call the reaper function on it, which will
--- dhcp-4.1-ESV-R8/client/dhclient.c.orig	2013-08-23 16:01:50.039299763 +0200
+++ dhcp-4.1-ESV-R8/client/dhclient.c	2013-08-23 16:24:56.087316151 +0200
@@ -98,6 +98,16 @@
 static int check_option_values(struct universe *universe, unsigned int opt,
 			       const char *ptr, size_t len);
 
+static void get_monotonic_time (struct timeval *tv)
+{
+  struct timespec tp;
+
+  clock_gettime (CLOCK_MONOTONIC, &tp);
+
+  tv->tv_sec  = tp.tv_sec;
+  tv->tv_usec = tp.tv_nsec / 1000;
+}
+
 int
 main(int argc, char **argv) {
 	int fd;
@@ -430,7 +440,7 @@
 	}
 
 	/* Get the current time... */
-	gettimeofday(&cur_tv, NULL);
+        get_monotonic_time(&cur_tv);
 
 	sockaddr_broadcast.sin_family = AF_INET;
 	sockaddr_broadcast.sin_port = remote_port;
@@ -3356,7 +3366,7 @@
 		client -> envc = 0;
 	}
 	dfree (envp, MDL);
-	gettimeofday(&cur_tv, NULL);
+        get_monotonic_time(&cur_tv);
 	return (WIFEXITED (wstatus) ?
 		WEXITSTATUS (wstatus) : -WTERMSIG (wstatus));
 }
--- dhcp-4.1-ESV-R8/server/dhcpd.c.orig	2013-01-03 00:36:27.000000000 +0100
+++ dhcp-4.1-ESV-R8/server/dhcpd.c	2013-08-23 16:30:57.931320429 +0200
@@ -227,6 +227,16 @@
 }
 #endif /* PARANOIA */
 
+static void get_monotonic_time (struct timeval *tv)
+{
+  struct timespec tp;
+
+  clock_gettime (CLOCK_MONOTONIC, &tp);
+
+  tv->tv_sec  = tp.tv_sec;
+  tv->tv_usec = tp.tv_nsec / 1000;
+}
+
 #ifndef UNIT_TEST
 int 
 main(int argc, char **argv) {
@@ -600,7 +610,7 @@
 	}
 
 	/* Get the current time... */
-	gettimeofday(&cur_tv, NULL);
+	get_monotonic_time(&cur_tv);
 
 	/* Set up the initial dhcp option universe. */
 	initialize_common_option_spaces ();
--- dhcp-4.1-ESV-R8/relay/dhcrelay.c.orig	2013-01-03 00:36:51.000000000 +0100
+++ dhcp-4.1-ESV-R8/relay/dhcrelay.c	2013-08-23 16:29:50.531319633 +0200
@@ -165,6 +165,16 @@
 	log_fatal(DHCRELAY_USAGE);
 }
 
+static void get_monotonic_time (struct timeval *tv)
+{
+  struct timespec tp;
+
+  clock_gettime (CLOCK_MONOTONIC, &tp);
+
+  tv->tv_sec  = tp.tv_sec;
+  tv->tv_usec = tp.tv_nsec / 1000;
+}
+
 int 
 main(int argc, char **argv) {
 	isc_result_t status;
@@ -524,7 +534,7 @@
 #endif
 
 	/* Get the current time... */
-	gettimeofday(&cur_tv, NULL);
+	get_monotonic_time(&cur_tv);
 
 	/* Discover all the network interfaces. */
 	discover_interfaces(DISCOVER_RELAY);
