--- dhcp-3.0.5-orig/client/clparse.c	2006-02-22 23:43:27.000000000 +0100
+++ dhcp-3.0.5/client/clparse.c	2010-12-10 10:46:12.000000000 +0100
@@ -82,7 +82,7 @@
 	top_level_config.bootp_policy = P_ACCEPT;
 	top_level_config.script_name = path_dhclient_script;
 	top_level_config.requested_options = default_requested_options;
-	top_level_config.omapi_port = -1;
+	top_level_config.omapi_port = 7979;
 	top_level_config.do_forward_update = 1;
 
 	group_allocate (&top_level_config.on_receipt, MDL);
--- dhcp-3.0.5-orig/omapip/listener.c	2004-06-10 19:59:48.000000000 +0200
+++ dhcp-3.0.5/omapip/listener.c	2010-12-10 14:58:07.000000000 +0100
@@ -51,14 +51,18 @@
 			   int max)
 {
 	omapi_addr_t addr;
+        struct in_addr localhost;
 
 #ifdef DEBUG_PROTOCOL
 	log_debug ("omapi_listen(port=%d, max=%d)", port, max);
 #endif
 
+        localhost.s_addr = htonl(INADDR_LOOPBACK);
+
 	addr.addrtype = AF_INET;
 	addr.addrlen = sizeof (struct in_addr);
 	memset (addr.address, 0, sizeof addr.address); /* INADDR_ANY */
+        memcpy (addr.address, &localhost, sizeof (struct in_addr));
 	addr.port = port;
 
 	return omapi_listen_addr (h, &addr, max);
