--- libmms-0.5/src/mmsx.c	2009-10-31 13:35:28.000000000 +0100
+++ libmms-0.5/src/mmsx.c	2010-06-29 11:16:50.000000000 +0200
@@ -42,14 +42,14 @@
   if (!mmsx)
     return mmsx;
     
-  mmsx->connection = mms_connect(io, data, url, bandwidth);
-  if (mmsx->connection)
-    return mmsx;
-    
   mmsx->connection_h = mmsh_connect(io, data, url, bandwidth);
   if (mmsx->connection_h)
     return mmsx;
 
+  mmsx->connection = mms_connect(io, data, url, bandwidth);
+  if (mmsx->connection)
+    return mmsx;
+
   free(mmsx);
   return NULL;
 }
