From ff20624b5815d0f86362f54237211a6771d4437b Mon Sep 17 00:00:00 2001
From: Sven Neumann <s.neumann@raumfeld.com>
Date: Tue, 12 Mar 2013 13:46:33 +0100
Subject: [PATCH] Plug a minor memory leak

Signed-off-by: Sven Neumann <s.neumann@raumfeld.com>
---
 libgupnp/gupnp-root-device.c |    1 +
 1 file changed, 1 insertion(+)

diff --git a/libgupnp/gupnp-root-device.c b/libgupnp/gupnp-root-device.c
index 2cdbad1..8b2f7ef 100644
--- a/libgupnp/gupnp-root-device.c
+++ b/libgupnp/gupnp-root-device.c
@@ -458,6 +458,7 @@ gupnp_root_device_constructor (GType                  type,
  DONE:
         /* Cleanup */
         g_free (location);
+        g_free (desc_path);
 
         return object;
 }
-- 
1.7.10.4

