--- gupnp-0.12.8-orig/libgupnp/gupnp-service-proxy.c	2009-05-11 13:48:39.000000000 +0200
+++ gupnp-0.12.8/libgupnp/gupnp-service-proxy.c	2009-11-16 12:10:12.000000000 +0100
@@ -1453,32 +1453,6 @@
                 return;
         }
 
-        /* We do not error out if proxy->priv->sid is NIL as the subscription
-         * response may not have been processed yet. */
-#if 0
-        if (!proxy->priv->sid) {
-                GUPnPContext *context;
-                GMainContext *main_context;
-
-                context = gupnp_service_info_get_context
-                                        (GUPNP_SERVICE_INFO (proxy));
-                main_context = gssdp_client_get_main_context
-                                        (GSSDP_CLIENT (context));
-
-                /* Wait. Perhaps the subscription response has not yet
-                 * been processed. */
-                g_main_context_iteration (main_context, FALSE);
-
-                if (!proxy->priv->sid || strcmp (hdr, proxy->priv->sid) != 0) {
-                        /* Really not our SID */
-                        soup_message_set_status
-                                (msg, SOUP_STATUS_PRECONDITION_FAILED);
-
-                        return;
-                }
-        }
-#endif
-
         hdr = soup_message_headers_get (msg->request_headers, "SEQ");
         if (hdr == NULL) {
                 /* No SEQ header */
