--- gupnp-0.12.8-orig/libgupnp//gupnp-service-proxy.c	2009-05-11 13:48:39.000000000 +0200
+++ gupnp-0.12.8/libgupnp/gupnp-service-proxy.c	2010-03-18 12:46:04.000000000 +0100
@@ -1731,6 +1731,13 @@
                                  msg->reason_phrase);
 
  ERROR:
+                /* Remove listener */
+                context = gupnp_service_info_get_context
+                                        (GUPNP_SERVICE_INFO (proxy));
+
+                server = gupnp_context_get_server (context);
+                soup_server_remove_handler (server, proxy->priv->path);
+
                 proxy->priv->subscribed = FALSE;
 
                 g_object_notify (G_OBJECT (proxy), "subscribed");
@@ -1742,13 +1749,6 @@
                                error);
 
                 g_error_free (error);
-
-                /* Remove listener */
-                context = gupnp_service_info_get_context
-                                        (GUPNP_SERVICE_INFO (proxy));
-
-                server = gupnp_context_get_server (context);
-                soup_server_remove_handler (server, proxy->priv->path);
         }
 }
 
