From d4c8a5ee1348493559ad6f30fa6164a62eb1fd0f Mon Sep 17 00:00:00 2001
From: Sven Neumann <s.neumann@raumfeld.com>
Date: Thu, 24 Jun 2010 13:11:27 +0200
Subject: [PATCH] Chain up in finalize()

Signed-off-by: Sven Neumann <s.neumann@raumfeld.com>
---
 libgupnp/gupnp-xml-doc.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/libgupnp/gupnp-xml-doc.c b/libgupnp/gupnp-xml-doc.c
index d1a9e0d..a85abab 100644
--- a/libgupnp/gupnp-xml-doc.c
+++ b/libgupnp/gupnp-xml-doc.c
@@ -52,6 +52,8 @@ gupnp_xml_doc_finalize (GObject *object)
         doc = GUPNP_XML_DOC (object);
 
         xmlFreeDoc (doc->doc);
+
+        G_OBJECT_CLASS (gupnp_xml_doc_parent_class)->finalize (object);
 }
 
 static void
-- 
1.7.0.4

