commit 40b60b57dcb49b57c2a34225395ef0761420f122
Author: Zeeshan Ali (Khattak) <zeeshanak@gnome.org>
Date:   Wed Jul 22 00:22:03 2009 +0300

    API to get SoupMessage associated with ServiceAction

diff --git a/doc/gupnp-sections.txt b/doc/gupnp-sections.txt
index 0ae3a3e..aae9e16 100644
--- a/doc/gupnp-sections.txt
+++ b/doc/gupnp-sections.txt
@@ -112,6 +112,7 @@ gupnp_service_action_set_valist
 gupnp_service_action_set_value
 gupnp_service_action_return
 gupnp_service_action_return_error
+gupnp_service_action_get_message
 gupnp_service_notify
 gupnp_service_notify_valist
 gupnp_service_notify_value
diff --git a/libgupnp/gupnp-service.c b/libgupnp/gupnp-service.c
index 9085015..d0f75bd 100644
--- a/libgupnp/gupnp-service.c
+++ b/libgupnp/gupnp-service.c
@@ -586,6 +586,21 @@ gupnp_service_action_return_error (GUPnPServiceAction *action,
         finalize_action (action);
 }
 
+/**
+ * gupnp_service_action_get_message
+ * @action: A #GUPnPServiceAction
+ *
+ * Get the #SoupMessage associated with @action. Mainly intended for
+ * applications to be able to read HTTP headers received from clients.
+ *
+ * Return #SoupMessage associated with @action. Unref after using it.
+ **/
+SoupMessage *
+gupnp_service_action_get_message  (GUPnPServiceAction *action)
+{
+        return g_object_ref (action->msg);
+}
+
 static void
 gupnp_service_init (GUPnPService *service)
 {
diff --git a/libgupnp/gupnp-service.h b/libgupnp/gupnp-service.h
index 0cdc33a..a6975b8 100644
--- a/libgupnp/gupnp-service.h
+++ b/libgupnp/gupnp-service.h
@@ -136,6 +136,9 @@ gupnp_service_action_return_error (GUPnPServiceAction *action,
                                    guint               error_code,
                                    const char         *error_description);
 
+SoupMessage *
+gupnp_service_action_get_message  (GUPnPServiceAction *action);
+
 void
 gupnp_service_notify              (GUPnPService *service,
                                    ...) G_GNUC_NULL_TERMINATED;
