--- ./libgupnp/gupnp-service-proxy.c	2009-05-11 13:48:39.000000000 +0200
+++ ./libgupnp/gupnp-service-proxy.c	2009-11-11 19:00:01.000000000 +0100
@@ -1675,6 +1675,16 @@
 
                 proxy->priv->sid = g_strdup (hdr);
 
+		/* Listen for events */
+		GUPnPContext *context = gupnp_service_info_get_context (GUPNP_SERVICE_INFO (proxy));
+		SoupServer *server = gupnp_context_get_server (context);
+
+		soup_server_add_handler (server,
+                                 proxy->priv->path,
+                                 server_handler,
+                                 proxy,
+                                 NULL);
+
                 /* Figure out when the subscription times out */
                 hdr = soup_message_headers_get (msg->response_headers,
                                                 "Timeout");
@@ -1761,7 +1771,6 @@
         GUPnPContext *context;
         SoupMessage *msg;
         SoupSession *session;
-        SoupServer *server;
         const char *server_url;
         char *sub_url, *delivery_url, *timeout;
 
@@ -1821,15 +1830,6 @@
                                      timeout);
         g_free (timeout);
 
-        /* Listen for events */
-        server = gupnp_context_get_server (context);
-
-        soup_server_add_handler (server,
-                                 proxy->priv->path,
-                                 server_handler,
-                                 proxy,
-                                 NULL);
-
         /* And send our subscription message off */
         proxy->priv->pending_messages =
                 g_list_prepend (proxy->priv->pending_messages, msg);
