From 341037cd1bee68cae56ce62d3eb03e7a8fe67d02 Mon Sep 17 00:00:00 2001
From: Ross Burton <ross@linux.intel.com>
Date: Wed, 12 Aug 2009 17:34:58 +0100
Subject: [PATCH] Handle the SOAPAction header being missing

---
 libgupnp/gupnp-service.c |    8 +++++++-
 1 files changed, 7 insertions(+), 1 deletions(-)

diff --git a/libgupnp/gupnp-service.c b/libgupnp/gupnp-service.c
index d0f75bd..d7b4c07 100644
--- a/libgupnp/gupnp-service.c
+++ b/libgupnp/gupnp-service.c
@@ -738,6 +738,11 @@ control_server_handler (SoupServer        *server,
         /* Get action name */
         soap_action = soup_message_headers_get (msg->request_headers,
                                                 "SOAPAction");
+        if (!soap_action) {
+                soup_message_set_status (msg, SOUP_STATUS_PRECONDITION_FAILED);
+                return;
+        }
+
         action_name = strchr (soap_action, '#');
         if (!action_name) {
                 soup_message_set_status (msg, SOUP_STATUS_PRECONDITION_FAILED);
@@ -751,7 +756,8 @@ control_server_handler (SoupServer        *server,
 
         /* This memory is libsoup-owned so we can do this */
         end = strrchr (action_name, '"');
-        *end = '\0';
+        if (end)
+                *end = '\0';
 
         /* Parse action_node */
         doc = xmlRecoverMemory (msg->request_body->data,
-- 
1.6.3.3

