--- a/libsoup/soup-session.c	2012-03-26 21:48:57.000000000 +0200
+++ b/libsoup/soup-session.c	2012-04-10 17:25:13.836175998 +0200
@@ -2310,6 +2310,7 @@
 	for (item = soup_message_queue_first (priv->queue);
 	     item;
 	     item = soup_message_queue_next (priv->queue, item)) {
+	        soup_session_unpause_message(session, item->msg);
 		soup_session_cancel_message (session, item->msg,
 					     SOUP_STATUS_CANCELLED);
 	}
