diff -Naur ./libsoup/soup-session.c ./libsoup-cp/soup-session.c
--- a/libsoup/soup-session.c	2014-12-07 15:55:22.000000000 +0100
+++ b/libsoup/soup-session.c	2015-06-02 13:49:51.437519717 +0200
@@ -2373,6 +2373,10 @@
 	g_return_if_fail (item != NULL);
 
 	item->paused = FALSE;
+
+	if (item->state == SOUP_MESSAGE_RUNNING)
+		soup_message_io_unpause (msg);
+	
 	soup_message_set_status (msg, status_code);
 	g_cancellable_cancel (item->cancellable);
 
