diff -ru gssdp-0.6.4.orig/libgssdp/gssdp-client.c gssdp-0.6.4/libgssdp/gssdp-client.c
--- gssdp-0.6.4.orig/libgssdp/gssdp-client.c	2009-01-06 15:44:15.000000000 +0100
+++ gssdp-0.6.4/libgssdp/gssdp-client.c	2009-11-04 20:26:12.000000000 +0100
@@ -642,3 +642,39 @@
 
         return socket_source_cb (client->priv->multicast_socket, client);
 }
+
+
+gboolean
+gssdp_client_reinit (GSSDPClient  *client)
+{
+
+        g_return_val_if_fail (GSSDP_IS_CLIENT (client), FALSE);
+
+        if (client->priv->multicast_socket) {
+                g_source_destroy ((GSource *) client->priv->multicast_socket);
+                client->priv->multicast_socket = NULL;
+        }
+        
+         client->priv->multicast_socket =
+                  gssdp_socket_source_new (GSSDP_SOCKET_SOURCE_TYPE_MULTICAST);
+        if (client->priv->multicast_socket != NULL) {
+                g_source_set_callback
+                        ((GSource *) client->priv->multicast_socket,
+                         multicast_socket_source_cb,
+                         client,
+                         NULL);
+
+                g_source_attach ((GSource *) client->priv->multicast_socket,
+                                 client->priv->main_context);
+                g_source_unref  ((GSource *) client->priv->multicast_socket);
+
+                return TRUE;
+        }
+   
+        g_printerr("%s: multicast socket still dead\n", G_STRFUNC);
+
+        return FALSE;
+}
+
+
+
diff -ru gssdp-0.6.4.orig/libgssdp/gssdp-client.h gssdp-0.6.4/libgssdp/gssdp-client.h
--- gssdp-0.6.4.orig/libgssdp/gssdp-client.h	2009-01-06 15:44:15.000000000 +0100
+++ gssdp-0.6.4/libgssdp/gssdp-client.h	2009-11-04 20:26:14.000000000 +0100
@@ -82,6 +82,10 @@
 const char *
 gssdp_client_get_server_id    (GSSDPClient  *client);
 
+gboolean
+gssdp_client_reinit        (GSSDPClient  *client);
+
+
 G_END_DECLS
 
 #endif /* __GSSDP_CLIENT_H__ */
diff -ru gssdp-0.6.4.orig/libgssdp/gssdp-resource-browser.c gssdp-0.6.4/libgssdp/gssdp-resource-browser.c
--- gssdp-0.6.4.orig/libgssdp/gssdp-resource-browser.c	2009-01-06 15:44:15.000000000 +0100
+++ gssdp-0.6.4/libgssdp/gssdp-resource-browser.c	2009-11-04 20:26:12.000000000 +0100
@@ -503,6 +503,10 @@
 {
         g_return_if_fail (GSSDP_IS_RESOURCE_BROWSER (resource_browser));
 
+        if(resource_browser->priv->client)
+                gssdp_client_reinit(resource_browser->priv->client);
+
+
         if (resource_browser->priv->active == active)
                 return;
 
